llvm/unittests/CAS/CASUtilsTest.cpp

```diff
@@ -0,0 +1,56 @@
+//===- CASUtilsTest.cpp ---------------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "llvm/CAS/CASDB.h"
+#include "llvm/CAS/Utils.h"
+#include "gtest/gtest.h"
+
+using namespace llvm;
+using namespace llvm::cas;
+
+TEST(CASUtilsTest, walkFileTreeRecursively) {
+  std::unique_ptr<CASDB> CAS = createInMemoryCAS();
+
+  auto make = [&](StringRef Content) {
+    return *expectedToOptional(CAS->createBlob(Content));
+  };
+
+  HierarchicalTreeBuilder Builder;
+  Builder.push(make("blob2"), TreeEntry::Regular, "/d2");
+  Builder.push(make("blob1"), TreeEntry::Regular, "/t1/d1");
+  Builder.push(make("blob3"), TreeEntry::Regular, "/t3/d3");
+  Builder.push(make("blob1"), TreeEntry::Regular, "/t3/t1nested/d1");
+  Optional<TreeRef> Root = expectedToOptional(Builder.create(*CAS));
+  ASSERT_TRUE(Root);
+
+  std::deque<std::pair<std::string, bool>> ExpectEntries{
+      {"/", true},
+      {"/d2", false},
+      {"/t1", true},
+      {"/t1/d1", false},
+      {"/t3", true},
+      {"/t3/d3", false},
+      {"/t3/t1nested", true},
+      {"/t3/t1nested/d1", false},
+  };
```

Comment:
I guess perf doesn't matter that much in unit tests, but `std::deque` seems like a heavyweight way to iterate through an array backwards. I was staring at the code for a while trying to figure out when you used push_back() on what I assumed was a queue.

I suggest using ArrayRef instead:
```
std::pair<std::string, bool> AllEntries[] = {..., {"/", true}};
auto RemainingEntries = makeArrayRef(AllEntries);
```
(then use `drop_front()`, `front()`, and `empty()` from ArrayRef)